client/lib/titan/new-mailbox.js

```diff
@@ -0,0 +1,245 @@
+	domain: valueIsNonEmpty,
+	isAdmin: valueIsBoolean,
+	mailbox: valueIsNonEmpty,
+	name: valueIsNonEmpty,
+	password: valueIsNonEmpty,
+};
+
+const getMailboxRequiredBooleanValue = () =>
+	PropTypes.shape( {
+		value: PropTypes.bool.isRequired,
+		error: PropTypes.string,
+	} );
+
+const getMailboxOptionalStringValue = () =>
+	PropTypes.shape( {
+		value: PropTypes.string.isRequired,
+		error: PropTypes.string,
+	} );
+
+const getMailboxRequiredStringValue = () =>
+	PropTypes.shape( {
+		value: PropTypes.string.isRequired,
+		error: PropTypes.string,
+	} ).isRequired;
+
+const getMailboxPropTypeShape = () =>
+	PropTypes.shape( {
+		uuid: PropTypes.string.isRequired,
+		alternativeEmail: getMailboxOptionalStringValue(),
+		domain: getMailboxRequiredStringValue(),
+		isAdmin: getMailboxRequiredBooleanValue(),
+		mailbox: getMailboxRequiredStringValue(),
+		name: getMailboxRequiredStringValue(),
+		password: getMailboxRequiredStringValue(),
+	} );
+
+const sanitizeEmailSuggestion = ( emailSuggestion ) =>
+	String( emailSuggestion )
+		.replace( /[^0-9a-z_'.-]/gi, '' )
+		.toLowerCase();
+
+const validateMailboxesAreUnique = ( mailboxes ) => {
+	const mailboxNameCounts = mailboxes.reduce( ( nameCount, mailbox ) => {
+		nameCount[ mailbox.mailbox ] = 1 + nameCount[ mailbox.mailbox ] ?? 0;
+		return nameCount;
+	}, {} );
+
+	return mailboxes.map( ( mailbox ) => {
+		if ( mailboxNameCounts[ mailbox.mailbox ] > 1 ) {
+			return {
+				...mailbox,
+				[ mailbox ]: { value: mailbox.mailbox, error: translate( 'Please use unique mailboxes' ) },
+			};
+		}
+		return mailbox;
+	} );
+};
+
+const mapMailboxFields = ( mailbox, callback, skipFields = [] ) => {
+	return Object.entries( mailbox )
+		.map( ( entry ) => {
+			const fieldName = entry[ 0 ];
+			if ( fieldName === 'uuid' || skipFields.includes( fieldName ) ) {
+				return entry;
+			}
+			const fieldValue = entry[ 1 ];
+			return [ fieldName, callback( fieldValue, fieldName ) ];
+		} )
+		.reduce( ( updatedMailbox, entry ) => {
+			updatedMailbox[ entry[ 0 ] ] = entry[ 1 ];
+			return updatedMailbox;
+		}, {} );
+};
+
+const removePreviousError = ( { value } ) => ( { value, error: null } );
+
+const clearPreviousErrors = ( mailboxes ) => {
+	return mailboxes.map( ( mailbox ) => mapMailboxFields( mailbox, removePreviousError ) );
+};
+
+const validateRequiredField = ( { value, error } ) => ( {
+	value,
+	error:
+		! error && typeof value !== 'boolean' && ( ! value || '' === value.trim() )
+			? translate( 'This field is required.' )
+			: error,
+} );
+
+const validateRequiredMailboxFields = ( mailbox, optionalFields = [] ) => {
+	return mapMailboxFields( mailbox, validateRequiredField, optionalFields );
+};
+
+const validateFullEmailAddress = ( { value, error }, allowEmpty = false ) => {
+	if ( error ) {
+		return { value, error };
+	}
+	if ( allowEmpty && value === '' ) {
+		return { value, error };
+	}
+	const invalidEmailAddressError = translate( 'Please supply a valid email address.' );
```

Comment:
Can we maybe re-use the same validator that G Suite uses here? https://github.com/Automattic/wp-calypso/blob/de5adf95b71694bc37170fd78006a73c5609c6c8/client/lib/gsuite/new-users.ts#L4

Reply:
I was really happy to use this! 😍 for the suggestion!